extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js

```diff
@@ -44,10 +48,31 @@ export class AmpStoryShoppingTag extends AMP.BaseElement {
 
     /** @param {!ShoppingConfigDataDef} tagData */
     this.tagData_ = null;
+
+    /** @param {?AmpElement} element */
+    this.shoppingAttachment_ = null;
+
+    /** @param {!AmpElement} element */
+    this.shoppingTagEl_ = null;
   }
 
   /** @override */
   buildCallback() {
+    /* This is used to prevent the shopping tag component from building if there is no shopping attachment. */
+    const pageElement = closestAncestorElementBySelector(
+      this.element,
+      'amp-story-page'
+    );
+
+    this.shoppingAttachment_ = childElementByTag(
```

Comment:
After you scope product ids per page (#37600), the `-attachment` element should be responsible of dispatching the data. If there's no `-attachment` element, then the `-tag` element should never consume any data. 

I suggest relying on this data effect instead. This way there would be no need insert a dependency on the tree shape like we're doing here.

Reply:
And since you're rendering only once, there's a nice lazy-instantiation pattern we could follow with `upgradeCallback`. This way, you can ensure that you have access to Services and the tag data right from the get-go. If there's no data, there's no upgrade.

I've demonstrated this on [#37711](https://github.com/ampproject/amphtml/pull/37711/files)

Reply:
I've realized that an `upgradeCallback` would be hard to test for this particular case. Feel free to disregard this!